File: CodingInterviewPatterns/1.Two_Pointers/4.Largest_Container/LargestContainer.java
/*

You are given an array of numbers, each representing the height of a vertical line on a graph.
A container can be formed with any pair of these lines, along with the x-axis of the graph. 
Return the amount of water which the largest container can hold.

Input: heights = [2, 7, 8, 3, 7, 6]
Output: 24

*/


import java.util.ArrayList;

public class Main {
    public int largest_container(ArrayList<Integer> heights) {

        // Delcare left and right pointer
        int left = 0;
        int right = heights.size() - 1;

        // Initial max water
        int maxWater = 0;

        // While left < right continue
        while (left < right) {

            // Get the current area (Base * height)
            int currArea = Math.min(heights.get(left),heights.get(right)) * (right - left);

            // Get max current area
            maxWater = Math.max(currArea,maxWater);

            // If left is < right increase left
            if (heights.get(left) < heights.get(right)){
                left++;
            // If left is > right decrease right
            } else if (heights.get(left) > heights.get(right)){
                right--;
            // Both are equal and move inwards both
            } else {
                right--;
                left++;
            }
        }

        // Return max mater
        return maxWater;

    }
}


// Time complexity: O(n)

File: CodingInterviewPatterns/1.Two_Pointers/4.Largest_Container/largest_container.py
'''

You are given an array of numbers, each representing the height of a vertical line on a graph. 
A container can be formed with any pair of these lines, along with the x-axis of the graph. 
Return the amount of water which the largest container can hold.

Input: heights = [2, 7, 8, 3, 7, 6]
Output: 24

'''


from typing import List

def largest_container(heights: List[int]) -> int:
    
    # Declare left and right pointer
    left, right = 0, len(heights) - 1

    # Max water
    max_water = 0

    # While left < right continue
    while left < right:

        # Get the current area (base * height)
        area = min(heights[left],heights[right]) * (right - left)

        # Update max current
        if area > max_water :
            max_water = area

        # If left is less move left one
        if heights[left] < heights[right]:
            left += 1
        
        # If left is greater decrease right
        elif heights[left] > heights[right]:
            right -= 1
        
        # If both are equal move both
        else:
            left += 1
            right -= 1
        
    # Return max area of water
    return max_water


# Time complexity: O(n)
